src/Kmd.Logic.Gateway.Automation/Publish.cs

```diff
@@ -102,22 +102,84 @@ private async Task CreateOrUpdateProducts(Guid subscriptionId, Guid providerId,
+                    ValidationStatus.CanBeUpdated => await this.UpdateProduct(client, subscriptionId, providerId, folderPath, product, productValidationResult.EntityId.Value).ConfigureAwait(false),
+                    _ => throw new NotSupportedException("Unsupported ValidationStatus in CreateOrUpdateProducts"),
+                };
+
+                await this.CreateOrUpdatePolicies(
+                    client: client,
+                    subscriptionId: subscriptionId,
+                    entityId: productId,
+                    folderPath: folderPath,
+                    policiesResults: productValidationResult.Policies,
+                    entityType: "Product",
+                    rateLimitPolicy: product.RateLimitPolicy,
+                    customPolicies: product.CustomPolicies).ConfigureAwait(false);
+            }
+        }
+
+        private async Task CreateOrUpdatePolicies(
+            IGatewayClient client,
+            Guid subscriptionId,
+            Guid? entityId,
+            string folderPath,
+            PoliciesValidationResult policiesResults,
+            string entityType,
+            RateLimitPolicy rateLimitPolicy,
+            IEnumerable<CustomPolicy> customPolicies)
+        {
+            if (entityId.HasValue)
+            {
+                if (rateLimitPolicy != null)
+                {
+                    var rateLimitPolicyRequest = new RateLimitPolicyRequest(
+                        rateLimitPolicy.Name, entityId, entityType, rateLimitPolicy.Description, rateLimitPolicy.Calls, rateLimitPolicy.RenewalPeriod);
+                    switch (policiesResults.RateLimitPolicy.Status)
+                    {
+                        case ValidationStatus.CanBeCreated:
+                            var created = await client.CreateRateLimitPolicyAsync(subscriptionId, rateLimitPolicyRequest).ConfigureAwait(false);
+                            this.publishResults.Add(new GatewayAutomationResult() { ResultCode = ResultCode.RateLimitPolicyCreated, EntityId = created.Id });
+                            break;
+                        case ValidationStatus.CanBeUpdated:
+                            // TODO: Update RateLimitPolicy
+                            break;
+                        default:
+                            throw new NotSupportedException("Unsupported RateLimitPolicy ValidationStatus in CreateOrUpdatePolicies");
+                    }
+                }
+
+                if (customPolicies != null)
+                {
+                    foreach (var customPolicy in customPolicies)
+                    {
+                        var customPolicyResult = policiesResults.CustomPolicies.Single(cp => string.Equals(cp.Name, customPolicy.Name, StringComparison.OrdinalIgnoreCase));
+
+                        using var xmlFs = new FileStream(path: Path.Combine(folderPath, customPolicy.XmlFile), FileMode.Open, FileAccess.Read);
```

Comment:
A subtle improvement chance. StreamReader has an overload that accepts the file path. Stream is not needed in that case

Reply:
<3 Done.